Source/DafnyCore/Resolver/PreType/PreTypeResolve.cs

```diff
@@ -480,42 +493,161 @@ private PreTypeResolver(ModuleResolver resolver, PreTypeInferenceModuleState pre
       Constraints.AddConfirmation(check, preType, tok, errorFormatString, onProxyAction);
     }
 
-    void AddComparableConstraint(PreType a, PreType b, IToken tok, string errorFormatString) {
-      Contract.Requires(a != null);
-      Contract.Requires(b != null);
-      Contract.Requires(tok != null);
-      Contract.Requires(errorFormatString != null);
-      Constraints.AddGuardedConstraint(() => ApplyComparableConstraints(a, b, tok, errorFormatString));
+    void AddComparableConstraint(PreType a, PreType b, IToken tok, bool forAsOrIs, string errorFormatString) {
+      // A "comparable types" constraint involves a disjunction. This can get gnarly for inference, so the full disjunction
+      // is checked post inference. The constraint can, however, be of use during inference, so we also add an approximate
+      // constraint (which is set up NOT to generate any error messages by itself, since otherwise errors would be duplicated).
+      Constraints.AddGuardedConstraint(() => ApproximateComparableConstraints(a, b, tok, forAsOrIs,
+        "(Duplicate error message) " + errorFormatString, false));
+      Constraints.AddConfirmation(tok, () => CheckComparableTypes(a, b, forAsOrIs), () => string.Format(errorFormatString, a, b));
+    }
+
+    /// <summary>
+    /// This method returns whether or not A and B are comparable types (notated with the constraint A ~~ B).
+    ///
+    /// The meaning of a comparable constraint
+    ///     A ~~ B
+    /// is the disjunction
+    ///     A ::> B    or    B ::> A
+    ///
+    /// If "!forAsOrIs", then "X ::> Y" means
+    ///     X :> Y
+    ///
+    /// If "forAsOrIs", then "X ::> Y" means
+    ///     X' :> Y', or
+    ///     X' and Y' are various bv types, or
+    ///     X' is int and Y' is in {int, char, bv, ORDINAL, real}.
+    /// where X' and Y' are the newtype ancestors of X and Y, respectively.
+    /// Additionally, under the legacy option /generalNewtypes:0 (which will be phased out over time), the latter also allows
+    /// several additional cases, see IsConversionCompatible.
+    /// </summary>
+    bool CheckComparableTypes(PreType a, PreType b, bool forAsOrIs) {
+      if (PreType.Same(a, b)) {
+        // this allows the case where "a" and "b" are proxies that are equal
+        return true;
+      }
+      if (a.Normalize() is not DPreType aa || b.Normalize() is not DPreType bb) {
+        return false;
+      }
+      if (IsSuperPreTypeOf(aa, bb) || IsSuperPreTypeOf(bb, aa)) {
+        return true;
+      }
+      if (!forAsOrIs) {
+        return false;
+      }
+      if (IsConversionCompatible(aa, bb) || IsConversionCompatible(bb, aa)) {
+        return true;
+      }
+      return false;
+    }
+
+    bool IsConversionCompatible(DPreType a, DPreType b) {
```

Comment:
The method is asymmetric, but the arguments names don't show that. Should they be `from` and `to` ?